### PR TITLE
v1alpha4 webhook

### PR DESCRIPTION
**Pull request by sadysnaat**
- [x] hub implementation

- [x] webhook tests

- [x] test cluster creation

### CODE REVIEW

**Comment by spectro-prow** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/spectrocloud/cluster-api-provider-maas/pull/18#" title="Author self-approved">sadysnaat</a>*
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please assign **jzhoucliqr**
You can assign the PR to them by writing `/assign @jzhoucliqr` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=spectrocloud%2Fcluster-api-provider-maas).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/spectrocloud/cluster-api-provider-maas/blob/main/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by spectro-prow** [comment]
Hi @sadysnaat. Thanks for your PR.

I'm waiting for a [spectrocloud](https://github.com/orgs/spectrocloud/people) member to verify that this patch is reasonable to test. If it is, they should reply with `/ok-to-test` on its own line. Until that is done, I will not automatically test new commits in this PR, but the usual testing commands by org members will still work. Regular contributors should [join the org](https://github.com/orgs/spectrocloud/people) to skip this step.

Once the patch is verified, the new status will be reflected by the `ok-to-test` label.

I understand the commands that are listed [here](https://go.k8s.io/bot-commands?repo=spectrocloud%2Fcluster-api-provider-maas).

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

**Comment by spectro-prow** [comment]
@sadysnaat: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
cluster-api-provider-maas_build | 718ddc489a61d808e6c130c5632df4c2d5631f37 | [link](http://34.120.121.77/view/gcs/capi-prow-artifacts/pr-logs/pull/spectrocloud_cluster-api-provider-maas/18/cluster-api-provider-maas_build/1428359617036947456) | `/test cluster-api-provider-maas_build`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by spectro-prow** [comment]
@sadysnaat: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
cluster-api-provider-maas_build | 902754787c7d93d0603191a37c5560582581e1e2 | [link](http://34.120.121.77/view/gcs/capi-prow-artifacts/pr-logs/pull/spectrocloud_cluster-api-provider-maas/18/cluster-api-provider-maas_build/1428368173010259968) | `/test cluster-api-provider-maas_build`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by spectro-prow** [comment]
@sadysnaat: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
cluster-api-provider-maas_build | 911306f1a9c1eb736423bc1271483e36f8753288 | [link](http://34.120.121.77/view/gcs/capi-prow-artifacts/pr-logs/pull/spectrocloud_cluster-api-provider-maas/18/cluster-api-provider-maas_build/1428374590102441984) | `/test cluster-api-provider-maas_build`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

File: api/v1alpha3/types.go
Comment (jzhoucliqr) [inline api/v1alpha3/types.go]:
```suggestion
// Instance describes an MAAS Machine.
```

File: api/v1alpha4/condition_consts.go
Comment (jzhoucliqr) [inline api/v1alpha4/condition_consts.go:2]: Is this file a direct copy from capi?

File: api/v1alpha4/condition_consts.go
Comment (jzhoucliqr) [inline api/v1alpha4/condition_consts.go]:
```suggestion
// Conditions and condition Reasons for the MaasMachine object
```

File: api/v1alpha4/maascluster_webhook.go
Comment (jzhoucliqr) [inline api/v1alpha4/maascluster_webhook.go]: any field we should prevent updating?

File: api/v1alpha4/maascluster_types.go
Comment (jzhoucliqr) [inline api/v1alpha4/maascluster_types.go:34]: maybe add some kubebuilder validation here? like length>1

File: api/v1alpha4/maasmachine_types.go
Comment (jzhoucliqr) [inline api/v1alpha4/maasmachine_types.go]: should it be maasmachine? or maascluster?

File: api/v1alpha4/maasmachine_types.go
Comment (jzhoucliqr) [inline api/v1alpha4/maasmachine_types.go]: please fix the description

File: api/v1alpha4/maasmachine_types.go
Comment (jzhoucliqr) [inline api/v1alpha4/maasmachine_types.go]: shall we use minMemoryInMB which is more clear?

File: api/v1alpha4/maasmachine_types.go
Comment (jzhoucliqr) [inline api/v1alpha4/maasmachine_types.go]: is this optional?

File: api/v1alpha4/maasmachine_types.go
Comment (jzhoucliqr) [inline api/v1alpha4/maasmachine_types.go:61]: add kubebuilder validation length

File: api/v1alpha4/maasmachine_types.go
Comment (jzhoucliqr) [inline api/v1alpha4/maasmachine_types.go:59]: fix description

File: api/v1alpha4/maasmachine_types.go
Comment (jzhoucliqr) [inline api/v1alpha4/maasmachine_types.go]: field name mis match json tag. any specific reason?

File: api/v1alpha4/maasmachine_webhook.go
Comment (jzhoucliqr) [inline api/v1alpha4/maasmachine_webhook.go]: please add validation for immutible fields.

File: api/v1alpha4/maasmachinetemplate_webhook.go
Comment (jzhoucliqr) [inline api/v1alpha4/maasmachinetemplate_webhook.go]: cleanup these comments.

**Comment by spectro-prow** [comment]
@sadysnaat: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
cluster-api-provider-maas_build | 911306f1a9c1eb736423bc1271483e36f8753288 | [link](http://34.120.121.77/view/gcs/capi-prow-artifacts/pr-logs/pull/spectrocloud_cluster-api-provider-maas/18/cluster-api-provider-maas_build/1428625496765108224) | `/test cluster-api-provider-maas_build`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

File: api/v1alpha3/types.go
Comment (sadysnaat) [reply api/v1alpha3/types.go]: Ack

File: api/v1alpha4/condition_consts.go
Comment (sadysnaat) [reply api/v1alpha4/condition_consts.go:2]: Yes

File: api/v1alpha4/condition_consts.go
Comment (sadysnaat) [reply api/v1alpha4/condition_consts.go]: Ack

**Comment by spectro-prow** [comment]
@sadysnaat: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
cluster-api-provider-maas_build | 50635b69080da77eb089422b3b99cf819f657ce6 | [link](http://34.120.121.77/view/gcs/capi-prow-artifacts/pr-logs/pull/spectrocloud_cluster-api-provider-maas/18/cluster-api-provider-maas_build/1428661610162950144) | `/test cluster-api-provider-maas_build`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

File: api/v1alpha4/maascluster_webhook.go
Comment (sadysnaat) [reply api/v1alpha4/maascluster_webhook.go]: Ack, added DNSDomain to it

File: api/v1alpha4/maascluster_types.go
Comment (sadysnaat) [reply api/v1alpha4/maascluster_types.go:34]: Ack

File: api/v1alpha4/maasmachine_types.go
Comment (sadysnaat) [reply api/v1alpha4/maasmachine_types.go]: Ack, changed

File: api/v1alpha4/maasmachine_types.go
Comment (sadysnaat) [reply api/v1alpha4/maasmachine_types.go]: Ack

File: api/v1alpha4/maasmachine_types.go
Comment (sadysnaat) [reply api/v1alpha4/maasmachine_types.go]: Ack, changed

File: api/v1alpha4/maasmachine_types.go
Comment (sadysnaat) [reply api/v1alpha4/maasmachine_types.go]: Ack changed to mendatory in v1alpah4

but now v1alpha3 -> v1alpha4 if these fields are not set have default values.

Refer. conversion.go

File: api/v1alpha4/maasmachine_types.go
Comment (sadysnaat) [reply api/v1alpha4/maasmachine_types.go:61]: Ack

File: api/v1alpha4/maasmachine_types.go
Comment (sadysnaat) [reply api/v1alpha4/maasmachine_types.go]: copy paste from previous files, changed

File: api/v1alpha4/maasmachinetemplate_webhook.go
Comment (sadysnaat) [reply api/v1alpha4/maasmachinetemplate_webhook.go]: Ack

File: api/v1alpha4/maasmachine_types.go
Comment (sadysnaat) [reply api/v1alpha4/maasmachine_types.go:59]: Ack

File: api/v1alpha4/maasmachine_webhook.go
Comment (sadysnaat) [reply api/v1alpha4/maasmachine_webhook.go]: Made image, minCPU and minMemory immutable for Machine and MachineTemplate

**Comment by spectro-prow** [comment]
@sadysnaat: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
cluster-api-provider-maas_build | ca8c3a20e5ecf832949a216da1ec30361a68e25a | [link](http://34.120.121.77/view/gcs/capi-prow-artifacts/pr-logs/pull/spectrocloud_cluster-api-provider-maas/18/cluster-api-provider-maas_build/1428695584679661568) | `/test cluster-api-provider-maas_build`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by spectro-prow** [comment]
@sadysnaat: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
cluster-api-provider-maas_build | ccc8e79e4077cdeafa07dec8ccc4a623208c38e5 | [link](http://34.120.121.77/view/gcs/capi-prow-artifacts/pr-logs/pull/spectrocloud_cluster-api-provider-maas/18/cluster-api-provider-maas_build/1428718989013422080) | `/test cluster-api-provider-maas_build`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by spectro-prow** [comment]
@sadysnaat: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
cluster-api-provider-maas_build | ccc8e79e4077cdeafa07dec8ccc4a623208c38e5 | [link](http://34.120.121.77/view/gcs/capi-prow-artifacts/pr-logs/pull/spectrocloud_cluster-api-provider-maas/18/cluster-api-provider-maas_build/1430376306280763392) | `/test cluster-api-provider-maas_build`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by spectro-prow** [comment]
@sadysnaat: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
cluster-api-provider-maas_build | ae1d0d0208f9be1186131ae7da0249afdf8708bc | [link](http://34.120.121.77/view/gcs/capi-prow-artifacts/pr-logs/pull/spectrocloud_cluster-api-provider-maas/18/cluster-api-provider-maas_build/1430410911117479936) | `/test cluster-api-provider-maas_build`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by spectro-prow** [comment]
@sadysnaat: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
cluster-api-provider-maas_build | 8b4d11d2bc7f01322cdfd4ed1d164d391f036930 | [link](http://34.120.121.77/view/gcs/capi-prow-artifacts/pr-logs/pull/spectrocloud_cluster-api-provider-maas/18/cluster-api-provider-maas_build/1430438467287912448) | `/test cluster-api-provider-maas_build`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

File: api/v1alpha3/conversion.go
Comment (jzhoucliqr) [inline api/v1alpha3/conversion.go]: whats the reason making this pointer vs non-pointer change?

Comment (sadysnaat) [reply api/v1alpha3/conversion.go]: https://github.com/spectrocloud/cluster-api-provider-maas/pull/18#discussion_r692362975 I thought we wanted to make these fields mendatory

Comment (sadysnaat) [reply api/v1alpha3/conversion.go]: Reverted back to pointers 

removed optional constraint

File: api/v1alpha3/maasmachine_types.go
Comment (jzhoucliqr) [inline api/v1alpha3/maasmachine_types.go:71]: will this json change cause issue for existing clusters?

File: api/v1alpha4/maasmachine_types.go
Comment (jzhoucliqr) [reply api/v1alpha4/maasmachine_types.go]: Either is fine, I just think maybe try to be consistent. 
if it's not optional, then should remove the omitempty tag
validation minimum, should it be 0 or 1 ?

File: api/v1alpha3/maasmachine_types.go
Comment (jzhoucliqr) [reply api/v1alpha3/maasmachine_types.go:71]: although it may not matter because this is in status.

Comment (sadysnaat) [reply api/v1alpha3/maasmachine_types.go:71]: In next reconcile it gets corrected

**Review by jzhoucliqr — APPROVED** [review]
/approve

**Comment by spectro-prow** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/spectrocloud/cluster-api-provider-maas/pull/18#pullrequestreview-741808146" title="Approved">jzhoucliqr</a>*, *<a href="https://github.com/spectrocloud/cluster-api-provider-maas/pull/18#" title="Author self-approved">sadysnaat</a>*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=spectrocloud%2Fcluster-api-provider-maas).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/spectrocloud/cluster-api-provider-maas/blob/main/OWNERS)~~ [jzhoucliqr]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Pull request merged**
